src/ejson2/EjsonDecoder.ts

```diff
@@ -0,0 +1,636 @@
+import {
+  BsonBinary,
+  BsonDbPointer,
+  BsonDecimal128,
+  BsonFloat,
+  BsonInt32,
+  BsonInt64,
+  BsonJavascriptCode,
+  BsonJavascriptCodeWithScope,
+  BsonMaxKey,
+  BsonMinKey,
+  BsonObjectId,
+  BsonSymbol,
+  BsonTimestamp,
+} from '../bson/values';
+import {Reader} from '@jsonjoy.com/util/lib/buffers/Reader';
+import {readKey} from '../json/JsonDecoder';
+import type {BinaryJsonDecoder} from '../types';
+
+export interface EjsonDecoderOptions {
+  /** Whether to parse legacy Extended JSON formats */
+  legacy?: boolean;
+}
+
+export class EjsonDecoder implements BinaryJsonDecoder {
```

Comment:
Make this class extends `JsonDecoder`. Reuse parent class methods, where it makes sense.